### PR TITLE
feat: add user notification preferences with configurable email, SMS, and push notification

### PR DESCRIPTION
**Pull request by Sarthak-btr**
### Description

This PR adds user notification preferences functionality, allowing users to configure their email, SMS, and push notification settings on a per-account basis. The implementation includes:

* A new database collection to store preferences
* REST API endpoints for managing preferences
* Integration with existing notification services to respect user choices

---

### New API Endpoints

1. **Update Notification Preferences**
   **PUT** `/accounts/<account_id>/notification-preferences`

   * **Request Body**:

     ```json
     {
       "email_enabled": boolean,
       "push_enabled": boolean,
       "sms_enabled": boolean
     }
     ```
   * **Response**:

     ```json
     {
       "email_enabled": boolean,
       "push_enabled": boolean,
       "sms_enabled": boolean
     }
     ```

2. **Get Account with Preferences**
   **GET** `/accounts/<id>?include_notification_preferences=true`

   * **Response**: Existing account object plus:

     ```json
     "notification_preferences": {
       "email_enabled": boolean,
       "push_enabled": boolean,
       "sms_enabled": boolean
     }
     ```

---

### Key Features

* **Default Preferences**: Automatically created (all enabled) when first accessed
* **Service Integration**: Notification services (email/SMS) accept optional `preferences` parameter and `bypass_preferences` flag
* **Validation & Error Handling**: Comprehensive input validation and descriptive error responses
* **Security**: Proper authentication and authorization checks to prevent cross-account access

---

#### New Collection: `account_notification_preferences`

```json
{
  "_id": ObjectId,
  "account_id": String (required, unique index),
  "email_enabled": Boolean (required, default: true),
  "push_enabled": Boolean (required, default: true),
  "sms_enabled": Boolean (required, default: true),
  "created_at": Date (required),
  "updated_at": Date (required)
}
```

* Added unique index on `account_id` field
* Enforced strict MongoDB schema validation
* Handled collection creation edge cases with appropriate errors

---

#### Automated Tests

**API Layer Tests** (`test_notification_preferences_api.py`):

* Get account with preferences included via query parameter
* Get account without preferences (default behavior)
* Update preferences with complete data
* Update preferences with partial/missing fields (defaults applied)
* Unauthorized access scenarios (no token, invalid token)
* Cross-account access prevention
* Disable all notification types

**Service Layer Tests** (`test_notification_preferences_service.py`):

* Automatic creation of default preferences when none exist
* Retrieval of existing preferences
* Creation of new preferences via update
* Updating existing preferences
* Disabling all notification types

---

#### Manual Test Cases

1. **Default Preferences Verification**

   * Create a new user via `POST /accounts`
   * Call `GET /accounts/{id}?include_notification_preferences=true`
   * Verify all preferences default to `true`

2. **Update Notification Preferences**

   * Authenticate as the user
   * Call `PUT /accounts/{account_id}/notification-preferences` with:

     ```json
     {"email_enabled": false, "sms_enabled": false}
     ```
   * Verify response shows updated preferences
   * Call `GET` endpoint to confirm persistence

3. **Notification Service Integration**

   * Disable email notifications for a user
   * Trigger an action that sends an email
   * Verify email is skipped and log message is generated
   * Test override using `bypass_preferences=true`

---

### Security Testing

* Attempt to update another user's preferences without proper auth → expect **401 Unauthorized**
* Test with expired/invalid tokens → verify rejection and error messages

---

### CODE REVIEW

**Comment by Sarthak-btr** [comment]
Default preferences are automatically set when creating new accounts:

    email_enabled: true
    sms_enabled: true
    push_enabled: true

File: src/apps/backend/modules/account/account_service.py [hunk 1]
```diff
@@ -61,3 +62,9 @@ def get_account_by_username(*, username: str) -> Account:
     @staticmethod
     def get_account_by_username_and_password(*, params: AccountSearchParams) -> Account:
         return AccountReader.get_account_by_username_and_password(params=params)
+
+    @staticmethod
+    def update_notification_preferences(*, params: UpdateNotificationPreferencesParams) -> Account:
+        AccountReader.get_account_by_id(params=AccountSearchByIdParams(id=params.account_id))
```
Comment (jjalan) [inline src/apps/backend/modules/account/account_service.py:68]: Why is this needed?

File: src/apps/backend/modules/account/internal/account_writer.py [hunk 2]
```diff
@@ -25,13 +27,17 @@ def create_account_by_username_and_password(*, params: CreateAccountByUsernameAn
...
+        default_notification_preferences = NotificationPreferences()
+
         account_bson = AccountModel(
             first_name=params.first_name,
             hashed_password=params_dict["hashed_password"],
             id=None,
             last_name=params.last_name,
             phone_number=None,
             username=params.username,
+            notification_preferences=default_notification_preferences,
```
Comment (jjalan) [inline src/apps/backend/modules/account/internal/account_writer.py:40]: nit: order arguments alphabetically

File: src/apps/backend/modules/account/internal/account_writer.py [hunk 3]
```diff
@@ -48,8 +54,17 @@ def create_account_by_phone_number(*, params: CreateAccountByPhoneNumberParams)
...
+
         account_bson = AccountModel(
-            first_name="", hashed_password="", id=None, last_name="", phone_number=phone_number, username=""
+            first_name="",
+            hashed_password="",
+            id=None,
+            last_name="",
+            phone_number=phone_number,
+            username="",
+            notification_preferences=default_notification_preferences,
```
Comment (jjalan) [inline src/apps/backend/modules/account/internal/account_writer.py:67]: nit: order arguments alphabetically

File: src/apps/backend/modules/account/internal/store/account_repository.py [hunk 4]
```diff
@@ -20,6 +20,15 @@
                 "description": "must be an object with country_code and phone_number",
             },
             "username": {"bsonType": "string", "description": "must be a string"},
+            "notification_preferences": {
```
Comment (jjalan) [inline src/apps/backend/modules/account/internal/store/account_repository.py:23]: Notification preferences should be stored in a separate collection - account_notification_preferences.

- With the current approach, if we add new functionality, it will make the account object fat which is not a scalable approach.

File: src/apps/backend/modules/notification/notification_service.py [hunk 5]
```diff
@@ -6,9 +11,47 @@
...
+                        message=f"SMS notification skipped: disabled by user preferences for account {account_id}"
+                    )
+                    return
+            except Exception as e:
+                Logger.error(message=f"Error checking notification preferences: {str(e)}")
+
         return SMSService.send_sms(params=params)
+
+    @staticmethod
+    def send_push_notification(*, message: str, account_id: str) -> None:
```
Comment (jjalan) [inline src/apps/backend/modules/notification/notification_service.py:44]: Lets add it once we have push notification

File: src/apps/backend/modules/notification/notification_service.py [hunk 6]
```diff
@@ -6,9 +11,47 @@
...
+        if account_id:
+            try:
+                account = AccountService.get_account_by_id(params=AccountSearchByIdParams(id=account_id))
+                if not account.notification_preferences or not account.notification_preferences.sms_enabled:
+                    Logger.info(
+                        message=f"SMS notification skipped: disabled by user preferences for account {account_id}"
+                    )
+                    return
+            except Exception as e:
+                Logger.error(message=f"Error checking notification preferences: {str(e)}")
```
Comment (jjalan) [inline src/apps/backend/modules/notification/notification_service.py:39]: When would exception be thrown?

File: src/apps/backend/modules/notification/notification_service.py [hunk 7]
```diff
@@ -6,9 +11,47 @@
...
+
         return EmailService.send_email(params=params)
 
     @staticmethod
-    def send_sms(*, params: SendSMSParams) -> None:
+    def send_sms(*, params: SendSMSParams, account_id: Optional[str] = None) -> None:
+        if account_id:
+            try:
+                account = AccountService.get_account_by_id(params=AccountSearchByIdParams(id=account_id))
```
Comment (jjalan) [inline src/apps/backend/modules/notification/notification_service.py:32]: Move this logic into sms/email service

File: src/apps/backend/modules/account/types.py [hunk 8]
```diff
@@ -61,6 +69,14 @@ class ResetPasswordParams:
     token: str
 
 
+@dataclass(frozen=True)
+class UpdateNotificationPreferencesParams:
+    account_id: str
+    email_enabled: bool
+    sms_enabled: bool
+    push_enabled: bool
```
Comment (jjalan) [inline src/apps/backend/modules/account/types.py:77]: nit: order properties alphabetically

File: src/apps/backend/modules/account/rest_api/notification_preferences_view.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+from modules.authentication.rest_api.access_auth_middleware import access_auth_middleware
+
+
+class NotificationPreferencesView(MethodView):
+    @access_auth_middleware
+    def get(self, account_id: str) -> ResponseReturnValue:
+        account = AccountService.get_account_by_id(params=AccountSearchByIdParams(id=account_id))
+
+        preferences = account.notification_preferences or NotificationPreferences()
+        return jsonify(asdict(preferences)), 200
```
Comment (jjalan) [inline src/apps/backend/modules/account/rest_api/notification_preferences_view.py:18]: Do we use `asdict` or create serializer?

File: src/apps/backend/modules/account/rest_api/account_router.py [hunk 10]
```diff
@@ -9,4 +10,11 @@ def create_route(*, blueprint: Blueprint) -> Blueprint:
         blueprint.add_url_rule("/accounts", view_func=AccountView.as_view("account_view"))
         blueprint.add_url_rule("/accounts/<id>", view_func=AccountView.as_view("account_view_by_id"), methods=["GET"])
         blueprint.add_url_rule("/accounts/<id>", view_func=AccountView.as_view("account_update"), methods=["PATCH"])
+
+        blueprint.add_url_rule(
+            "/accounts/<account_id>/notification-preferences",
```
Comment (jjalan) [inline src/apps/backend/modules/account/rest_api/account_router.py:15]: I am not sure if this is strictly needed, but rather should be sent when account object is serialized.

Perhaps, add `include_notification_preference` in query string when an account is fetched, and if its true, send notification preference down with it.

Its ok to keep put.

File: src/apps/backend/modules/account/internal/account_writer.py (hunk 2 above)
Comment (Sarthak-btr) [reply src/apps/backend/modules/account/internal/account_writer.py]: Updated - [commit](https://github.com/jalantechnologies/flask-react-template/pull/239/commits/4ad066a376192aa7c79474320b69226e81b6e618)

File: src/apps/backend/modules/account/internal/account_writer.py (hunk 3 above)
Comment (Sarthak-btr) [reply src/apps/backend/modules/account/internal/account_writer.py]: Updated - [commit](https://github.com/jalantechnologies/flask-react-template/pull/239/commits/4ad066a376192aa7c79474320b69226e81b6e618)

File: src/apps/backend/modules/account/types.py (hunk 8 above)
Comment (Sarthak-btr) [reply src/apps/backend/modules/account/types.py:76]: Updated - [commit](https://github.com/jalantechnologies/flask-react-template/pull/239/commits/e2db5839f065916fb2935ef7c7fb0b9de3e4278b)

File: src/apps/backend/modules/account/account_service.py [hunk 11]
```diff
@@ -1,13 +1,23 @@
+from datetime import datetime
+
+from pymongo import ReturnDocument
```
Comment (jjalan) [inline src/apps/backend/modules/account/account_service.py:3]: It is not Ok for service layer to know the underlying storage. Only reader and write should know that, and they should return data objects in return.

File: src/apps/backend/modules/account/internal/store/account_repository.py (hunk 4 above)
Comment (Sarthak-btr) [reply src/apps/backend/modules/account/internal/store/account_repository.py]: [refactor](https://github.com/jalantechnologies/flask-react-template/pull/239/commits/736de6536a5024b5ac9a0b79fa4161373c92afa5): move notification preferences to separate collection account_notification_preferences to improve scalability

File: src/apps/backend/modules/notification/notification_service.py (hunk 7 above)
Comment (Sarthak-btr) [reply src/apps/backend/modules/notification/notification_service.py]: Updated

File: src/apps/backend/modules/account/rest_api/account_router.py (hunk 10 above)
Comment (Sarthak-btr) [reply src/apps/backend/modules/account/rest_api/account_router.py:15]: Updated With The:
{{baseUrl}}/api/accounts/{{account_id}}?include_notification_preferences=true

File: src/apps/backend/modules/account/account_service.py [hunk 12]
```diff
@@ -61,3 +69,54 @@ def get_account_by_username(*, username: str) -> Account:
...
+
+        preferences_model = AccountNotificationPreferencesModel.from_bson(notification_preferences)
+        return NotificationPreferences(
+            email_enabled=preferences_model.email_enabled,
+            push_enabled=preferences_model.push_enabled,
+            sms_enabled=preferences_model.sms_enabled,
+        )
+
+    @staticmethod
+    def update_notification_preferences(params: UpdateNotificationPreferencesParams) -> NotificationPreferences:
```
Comment (Sarthak-btr) [inline src/apps/backend/modules/account/account_service.py:93]: For updating the notification preferences. Using for the route 
blueprint.add_url_rule(
            "/accounts/<account_id>/notification-preferences",
            view_func=NotificationPreferencesView.as_view("notification_preferences_update"),
            methods=["PUT"],
        )

File: src/apps/backend/modules/account/account_service.py (hunk 1 above)
Comment (Sarthak-btr) [reply src/apps/backend/modules/account/account_service.py]: For updating the notification preferences. Using for the route
blueprint.add_url_rule(
"/accounts/<account_id>/notification-preferences",
view_func=NotificationPreferencesView.as_view("notification_preferences_update"),
methods=["PUT"],
)

File: src/apps/backend/modules/account/types.py [hunk 13]
```diff
@@ -44,6 +44,13 @@ class AccountInfo:
     username: str
 
 
+@dataclass(frozen=True)
+class NotificationPreferences:
```
[the remark this answers is not in the document]
Comment (Sarthak-btr) [reply src/apps/backend/modules/account/types.py]: Updated. - [commit ](https://github.com/jalantechnologies/flask-react-template/pull/239/commits/f645a408cd625d3674b38bff0646589b3eeadfd5)

File: src/apps/backend/modules/notification/notification_service.py [hunk 14]
```diff
@@ -9,6 +12,18 @@ class NotificationService:
...
+        *, params: SendEmailParams, preferences: Optional[NotificationPreferences] = None
+    ) -> None:
+        return EmailService.send_email_with_preferences(params=params, preferences=preferences)
+
     @staticmethod
     def send_sms(*, params: SendSMSParams) -> None:
         return SMSService.send_sms(params=params)
+
+    @staticmethod
+    def send_sms_with_preferences(
```
[the remark this answers is not in the document]
Comment (Sarthak-btr) [reply src/apps/backend/modules/notification/notification_service.py]: updated `send_sms`

**Comment by Sarthak-btr** [comment]
move to - [PR](https://github.com/jalantechnologies/flask-react-template/pull/258)

**Comment by Sarthak-btr** [comment]
move to - https://github.com/jalantechnologies/flask-react-template/pull/258
